src/Enzo/enzo_EnzoMethodPmDeposit.cpp

```diff
@@ -165,172 +191,187 @@ void EnzoMethodPmDeposit::compute ( Block * block) throw()
-          const int dv =  particle.stride(it,ia_vx);
+      	  const int dp =  particle.stride(it,ia_x);
+      	  const int dv =  particle.stride(it,ia_vx);
+
+#ifdef DEBUG_COLLAPSE
+          CkPrintf ("DEBUG_COLLAPSE vxa[0] = %lg\n",vxa[0]);
+#endif
 
-          for (int ip=0; ip<np; ip++) {
+	        for (int ip=0; ip<np; ip++) {
 
-            double x = xa[ip*dp] + vxa[ip*dv]*dt;
+      	    double x = xa[ip*dp] + vxa[ip*dv]*dt;
 
-            double tx = nx*(x - xm) / (xp - xm) - 0.5;
+      	    double tx = nx*(x - xm) / (xp - xm) - 0.5;
 
-            int ix0 = gx + floor(tx);
+        	  int ix0 = gx + floor(tx);
 
-            int ix1 = ix0 + 1;
+      	    int ix1 = ix0 + 1;
 
-            double x0 = 1.0 - (tx - floor(tx));
-            double x1 = 1.0 - x0;
+      	    double x0 = 1.0 - (tx - floor(tx));
+      	    double x1 = 1.0 - x0;
 
-            de_p[ix0] += dens*x0;
-            de_p[ix1] += dens*x1;
+      	    de_p[ix0] += pdens[ip]*dens*x0;
+      	    de_p[ix1] += pdens[ip]*dens*x1;
 
-          }
+      	  } // np
 
         } else if (rank == 2) {
 
-          const int ia_x  = particle.attribute_index(it,"x");
-          const int ia_y  = particle.attribute_index(it,"y");
-          const int ia_vx = particle.attribute_index(it,"vx");
-          const int ia_vy = particle.attribute_index(it,"vy");
+      	  const int ia_x  = particle.attribute_index(it,"x");
+      	  const int ia_y  = particle.attribute_index(it,"y");
+      	  const int ia_vx = particle.attribute_index(it,"vx");
+      	  const int ia_vy = particle.attribute_index(it,"vy");
+
+      	  // Batch arrays
+      	  enzo_float * xa  = (enzo_float *)particle.attribute_array (it,ia_x,ib);
+      	  enzo_float * ya  = (enzo_float *)particle.attribute_array (it,ia_y,ib);
+      	  enzo_float * vxa = (enzo_float *)particle.attribute_array (it,ia_vx,ib);
+      	  enzo_float * vya = (enzo_float *)particle.attribute_array (it,ia_vy,ib);
+
+      	  const int dp =  particle.stride(it,ia_x);
+      	  const int dv =  particle.stride(it,ia_vx);
+
+	        for (int ip=0; ip<np; ip++) {
+
+      	    double x = xa[ip*dp] + vxa[ip*dv]*dt;
+      	    double y = ya[ip*dp] + vya[ip*dv]*dt;
+
+      	    double tx = nx*(x - xm) / (xp - xm) - 0.5;
+      	    double ty = ny*(y - ym) / (yp - ym) - 0.5;
+
+      	    int ix0 = gx + floor(tx);
+      	    int iy0 = gy + floor(ty);
+
+      	    int ix1 = ix0 + 1;
+      	    int iy1 = iy0 + 1;
+
+      	    double x0 = 1.0 - (tx - floor(tx));
+      	    double y0 = 1.0 - (ty - floor(ty));
+
+      	    double x1 = 1.0 - x0;
+      	    double y1 = 1.0 - y0;
+
+	          if ( dens < 0.0) {
+	            CkPrintf ("%s:%d ERROR: dens = %f\n", __FILE__,__LINE__,dens);
+	          }
+
+      	    de_p[ix0+mx*iy0] += pdens[ip]*dens*x0*y0;
+      	    de_p[ix1+mx*iy0] += pdens[ip]*dens*x1*y0;
+      	    de_p[ix0+mx*iy1] += pdens[ip]*dens*x0*y1;
+      	    de_p[ix1+mx*iy1] += pdens[ip]*dens*x1*y1;
```

Comment:
This doesn't make much sense to me. It should be something like 'density += pmass * inv_vol * x0 * y0'. The thing that needs to be checked is whether densities and volumes are comoving.

Reply:
Good question. This bit of the code I just mimicked from what was already there.... @jwise77 since you looked do you have insight on this (since I'm a bit tight on time I can spend diving into this)

Reply:
Again, this is related to my other comments. I bring this up because in a problem I was running I was creating star particles, with masses that had actual dimensions of mass, the code was crashing since the grid densities calculated here were completely wrong. The way I fixed it was to put in an 'inv_vol' term. However, I only tested it in the case of a unigrid with no cosmology. As I said in my other comments, the issue of particle masses vs densities needs to be cleared up in my opinion.

Reply:
In our last meeting, we agreed that this is an important point to raise for particles in Enzo-E. I don't think this should hold up accepting this PR. I've opened up issue #83 for a discussion there.